### DISCUSSION TITLE
Leverage OpenAPI Spec for Client/Model Generation?

### DISCUSSION THREAD

**Discussion by JerrettDavis** [awaescher/OllamaSharp#44] [General]
Today, the OllamaSharp library is largely hand-rolled. This was done because the [Ollama team](https://github.com/ollama/ollama) does not furnish any OpenAPI or Swagger specifications. There is an [issue](https://github.com/ollama/ollama/issues/3383) open in the Ollama repository to try to remedy this. The Ollama contributors have [stated a few times](https://github.com/ollama/ollama/issues/3487#issuecomment-2111872856) that they're not fans of the existing tooling out there to automatically generate OpenAPI specs from. 

I had started working [on my own OpenAPI spec](https://gist.github.com/JerrettDavis/5469a17c48849b6c60fa51cb17e9c124). It's not fully complete or accurate, but it's close. There's also the spec furnished by [@HavenDV](https://github.com/HavenDV) [here](https://github.com/tryAGI/Ollama/blob/main/docs/openapi.yaml).

HavenDV [has asked](https://github.com/awaescher/OllamaSharp/pull/40#issuecomment-2150527580) that we combine efforts a bit to create a standard ollama wrapper. I'm inclined to agree.

Ultimately, I'd like to use something like OllamaSharp or HavenDV's Ollama nuget package to underpin broader packages, like a Semantic Kernel connector. There are some out there today, but they either utilize older versions of OllamaSharp or they're also using hand-rolled implementations. 

I'm using some cycles to try to get the OpenAPI spec fully in-line with the current Ollama code. Further, I'm going to push hard to get the Ollama team to include the spec in the repository for ongoing releases. I'm working to script out and automate as much of this as possible to at least throw some CI warnings if the server API and OpenAPI spec start to drift. 

@awaescher What are your thoughts?

**Comment by awaescher** [comment]
First of all, I love this opener with the efforts to add all the links to further information, just checked them. And yes, I am totally in to get something like this rolling. I guess I need to say it here but personally I think it's obvious though: I don't need OllamaSharp to be hand-rolled. I am not voting for the current code base. The intention has always been to serve the most convenient way (maybe highly opinionated) to use the Ollama API from an .NET app - mostly because I think it's a good way to host the API separately from the application host. Most Ollama libraries required Ollama to be running locally.

That being said, I am not sure if the OpenAI spec can cover all the required endpoints to build powerfull Ollama API apps, like pulling models and getting the currencly running models, etc. So my first question would be: Would this make an OpenAI spec library + Ollama specifics? Or are there no specifics left?

Also, if you can manage to get the Ollama team to host the specs, I think we could link it from their repository to ours with a git submodule directly, for example.

**Comment by dcostea** [comment]
I'm a little bit confused, is this about **OpenAPI** Spec or about **OpenAI** Spec?

**Reply by JerrettDavis** [reply]
Actually a tiny bit of both! Ollama has limited [OpenAI API](https://github.com/ollama/ollama/blob/c39761c5525132d96e1da0956a9aa39e87b54114/server/routes.go#L989) support, but there are a couple more pull requests adding more OpenAI API interop. [1](https://github.com/ollama/ollama/pull/5007). [2](https://github.com/ollama/ollama/pull/5028)

**Comment by awaescher** [comment]
Wow man, this was my typo at first but it seems I got down the same rabbit hole while writing. It's OpenAPI of course 🙈

**Reply by dcostea** [reply]
> Wow man, this was my typo at first but it seems I got down the same rabbit hole while writing. It's OpenAPI of course 🙈
Thanks.

It happened to me more times when doing live speeches too!

**Comment by JerrettDavis** [comment]
I went ahead and opened a [PR on the Ollama repo](https://github.com/ollama/ollama/pull/5040) with the OpenAPI 3.1 spec.

**Reply by awaescher** [reply]
Let's see if they review a file THIS big!

**Reply by JerrettDavis** [reply]
I'm hopeful! I went through and tried to word-for-word move over most of their API documentation.
